### PR TITLE
Update ltoworkarounds.conf

### PR DESCRIPTION
**Pull request by Althorion**
I also need `ltofat` for `app-editors/vim-8.0.0386::gentoo  USE="X acl gpm lua luajit nls perl python -cscope -debug -minimal -racket -ruby (-selinux) -tcl -vim-pager" PYTHON_TARGETS="python2_7 python3_4 (-python3_5) (-python3_6)"`, but since you seem to be able to emerge it without problems, I decided against adding it.

### CODE REVIEW

**Comment by InBetweenNames** [comment]
Did you set up the ltoplugin symlink as suggested in the README.md?  I'll test out these locally as well.  Also, have you tried binutils 2.29?

**Comment by Althorion** [comment]
Yes, I’ve made that symlink and no, I haven’t tried newer binutils—I’m currently also trying to debug XCOM 2 errors and as all Feral ports, it is sensitive to binutils changes.

**Comment by InBetweenNames** [comment]
Really??  That's so odd--I'd understand if it were sensitive to glibc, but to be sensitive to binutils is truly an odd one.  One more thing to try as well: switch from ld.bfd to ld.gold, or vice versa.  I'm currently emerging these packages to see if I can repro these on my system

**Comment by Althorion** [comment]
I believe they do something ugly with linking, hence the sensitivity to binutils — possibly for the DRM protection. Glibc also can be a problem, though.

Still, changing the linker makes no difference, vim won’t build without `ltofat` and will build with it using both gold and BFD.

**Comment by InBetweenNames** [comment]
Just tested these workarounds--looks good to me, thank you very much for the contribution!  I just have a minor nitpick: would you mind condensing these two commits into one commit with a `git rebase -i HEAD~2`? Also, make sure you are using the new "lto" subdirectory--it should say "lto/nolto.conf".  Also, I see a bug in what I committed earlier so I'm going to go fix that.

With respect to the vim, I'm starting to think that the binutils are actually making a difference.  I'll try downgrading and seeing if I can repro that particular part.  I would like to try to isolate that issue you are having so we can get a lasting fix for it.

**Comment by Althorion** [comment]
Done. Making another PR now, still without Vim changes.

**Comment by InBetweenNames** [comment]
Yeah I just downgraded to 2.28.1 and I still don't need any workarounds for vim.  I'm wondering if something vim depends on in your system was not built with LTO?  Also, could you show me the symlink for your liblto_plugin.so?  Assuming you're on amd64, you should be able to get it with:

~~~
ls -l /usr/x86_64-pc-linux-gnu/binutils-bin/lib/bfd-plugins/liblto_plugin.so
~~~

Also your GCC version too please

**Comment by Althorion** [comment]
```
$ equery d vim
 * These packages depend on vim:
app-vim/gentoo-syntax-20170225 (>=app-editors/vim-7.3)
dev-util/ninja-1.7.2 (vim-syntax ? app-editors/vim)
virtual/editor-0 (app-editors/vim)
virtual/pager-0 (app-editors/vim[vim-pager])

$ ls -l /usr/x86_64-pc-linux-gnu/binutils-bin/lib/bfd-plugins/liblto_plugin.so
lrwxrwxrwx 1 root root 59 09-14 18:24 /usr/x86_64-pc-linux-gnu/binutils-bin/lib/bfd-plugins/liblto_plugin.so -> /usr/libexec/gcc/x86_64-pc-linux-gnu/7.2.0/liblto_plugin.so

$ gcc --version
gcc (Gentoo 7.2.0 p1.1) 7.2.0
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Since I did full `emerge -e @world`, I highly doubt anything could possibly depend on non-lto Vim.

**Comment by InBetweenNames** [comment]
Everything there looks good...could you post your make.conf as well?

**Comment by Althorion** [comment]
```
FLTO="-flto=8"
GRAPHITE="-fgraphite-identity -floop-interchange -ftree-loop-distribution -floop-strip-mine -floop-block"
AR=gcc-ar
NM=gcc-nm
RANLIB=gcc-ranlib

CFLAGS="-march=native -O3 -pipe ${GRAPHITE} ${FLTO} -fuse-linker-plugin"
#CFLAGS="-march=native -O2 -pipe"
CXXFLAGS="${CFLAGS}"
LDFLAGS="-Wl,--hash-style=gnu ${CFLAGS}"
MAKEOPTS="-j8"

CHOST="x86_64-pc-linux-gnu"

# ACCEPT_KEYWORDS="~amd64"
CPU_FLAGS_X86="aes avx f16c fma3 fma4 mmx mmxext popcnt sse sse2 sse3 sse4_1 sse4_2 sse4a ssse3 xop"
USE="${CPU_FLAGS_X86} -bindist -bluetooth -cdda -cups -dvd -libav -lirc -wayland -wifi acl alsa branding harfbuzz infinality kerberos lto mmx networkmanager pch pgo pulseaudio sse sse2 vaapi vdpau zsh-completion"
XFCE_PLUGINS=""
LINGUAS="en_GB en_US en_CA pl_PL en pl"
L10N="en-GB en-US en-CA pl-PL en pl"
ACCEPT_LICENSE="*"
INPUT_DEVICES="libinput"
VIDEO_CARDS="amdgpu radeonsi"
GRUB_PLATFORMS="efi-64"
# FONT_TYPES="otf"

PYTHON_TARGETS="python2_7 python3_4"
PYTHON_SINGLE_TARGET="python3_4"

PORTDIR="/usr/portage"
DISTDIR="${PORTDIR}/distfiles"
PKGDIR="${PORTDIR}/packages"

PORTAGE_ELOG_CLASSES="info war error log qa"
PORTAGE_ELOG_SYSTEM="save"
PORTAGE_NICENESS=15
```

**Comment by InBetweenNames** [comment]
Wow--I'm genuinely stumped!  One last thing: could you post the build.log of vim when you don't ltofat.conf it?  I want to see what kind of unresolved symbols are involved.

**Comment by Althorion** [comment]
Full log: https://gist.github.com/096cf66550a875d294edb242ff68c4eb
tl;dr;
```objects/arabic.o: plugin needed to handle lto object
objects/buffer.o: plugin needed to handle lto object
objects/blowfish.o: plugin needed to handle lto object
objects/crypt.o: plugin needed to handle lto object
objects/crypt_zip.o: plugin needed to handle lto object
objects/dict.o: plugin needed to handle lto object
objects/diff.o: plugin needed to handle lto object
objects/digraph.o: plugin needed to handle lto object
objects/edit.o: plugin needed to handle lto object
objects/eval.o: plugin needed to handle lto object
objects/evalfunc.o: plugin needed to handle lto object
objects/ex_cmds.o: plugin needed to handle lto object
objects/ex_cmds2.o: plugin needed to handle lto object
objects/ex_docmd.o: plugin needed to handle lto object
objects/ex_eval.o: plugin needed to handle lto object
objects/ex_getln.o: plugin needed to handle lto object
objects/farsi.o: plugin needed to handle lto object
objects/fileio.o: plugin needed to handle lto object
objects/fold.o: plugin needed to handle lto object
objects/getchar.o: plugin needed to handle lto object
objects/hardcopy.o: plugin needed to handle lto object
objects/hashtab.o: plugin needed to handle lto object
objects/if_cscope.o: plugin needed to handle lto object
objects/if_xcmdsrv.o: plugin needed to handle lto object
objects/list.o: plugin needed to handle lto object
objects/mark.o: plugin needed to handle lto object
objects/memline.o: plugin needed to handle lto object
objects/menu.o: plugin needed to handle lto object
objects/misc1.o: plugin needed to handle lto object
objects/misc2.o: plugin needed to handle lto object
objects/move.o: plugin needed to handle lto object
objects/mbyte.o: plugin needed to handle lto object
objects/normal.o: plugin needed to handle lto object
objects/ops.o: plugin needed to handle lto object
objects/option.o: plugin needed to handle lto object
objects/os_unix.o: plugin needed to handle lto object
objects/pathdef.o: plugin needed to handle lto object
objects/popupmnu.o: plugin needed to handle lto object
objects/quickfix.o: plugin needed to handle lto object
objects/regexp.o: plugin needed to handle lto object
objects/screen.o: plugin needed to handle lto object
objects/search.o: plugin needed to handle lto object
objects/sha256.o: plugin needed to handle lto object
objects/spell.o: plugin needed to handle lto object
objects/spellfile.o: plugin needed to handle lto object
objects/syntax.o: plugin needed to handle lto object
objects/tag.o: plugin needed to handle lto object
objects/term.o: plugin needed to handle lto object
objects/ui.o: plugin needed to handle lto object
objects/undo.o: plugin needed to handle lto object
objects/userfunc.o: plugin needed to handle lto object
objects/version.o: plugin needed to handle lto object
objects/window.o: plugin needed to handle lto object
objects/if_lua.o: plugin needed to handle lto object
objects/if_perl.o: plugin needed to handle lto object
objects/if_perlsfio.o: plugin needed to handle lto object
objects/if_python.o: plugin needed to handle lto object
objects/if_python3.o: plugin needed to handle lto object
objects/netbeans.o: plugin needed to handle lto object
objects/channel.o: plugin needed to handle lto object
objects/charset.o: plugin needed to handle lto object
objects/json.o: plugin needed to handle lto object
objects/main.o: plugin needed to handle lto object
objects/memfile.o: plugin needed to handle lto object
objects/message.o: plugin needed to handle lto object
```


Another though—do our vim versions (I use stable, 8.0.0386) and USE flags match?

**Comment by InBetweenNames** [comment]
Oh definitely not--I'm on the ~amd64 vim not the stable one.  Our USE flags are different too--would you try upgrading to the testing one?

**Comment by InBetweenNames** [comment]
`plugin needed to handle lto object` is usually something that happens if -fuse-linker-plugin isn't used.  But your make.conf does seem to have it.  Very interesting.  I know some people had that problem if the symlink wasn't present, but this is a first for me.

**Comment by Althorion** [comment]
Done, but no dice—same errors. I have also check the ebuild for anything that would disable `-fuse-linker-plugin`, but find nothing of sort.

**Comment by preney** [comment]
I am able to try migrating a system to LTO w/graphite and -O3. It currently is stable. I am thinking, except for GCC v7.2.0, perhaps I should keep it stable and see what happens (e.g., will vim build), etc.?

**Comment by InBetweenNames** [comment]
Yeah @preney that would be a good exercise to see if we can repro that particular problem. @Althorion would you be willing to try binutils 2.29 temporarily just to see if the problem still happens with that?  You could revert after you test it out--not an intrusive change.

**Comment by InBetweenNames** [comment]
@Althorion Checking the build log, the linker appears to have an fno-lto in its arguments.  This could be the cause.

**Comment by InBetweenNames** [comment]
Also vim is being built with O2 but your make.conf specifies O3. Do you have any other env overrides other than the LTO ones?

**Comment by Althorion** [comment]
Vim is built with `-O2`, because its ebuild enforces it: [see line 147](https://gitweb.gentoo.org/repo/gentoo.git/tree/app-editors/vim/vim-8.0.1056.ebuild#n147).
I have no clue where did the `-fno-lto` came from, though—I have no other env overrides:
```
$ fgrep -r -i vim /etc/portage
/etc/portage/package.use/vim:app-editors/vim lua luajit perl python
/etc/portage/package.accept_keywords/vim:app-editors/vim
/etc/portage/package.accept_keywords/vim:app-editors/vim-core
/etc/portage/package.env/ltoworkarounds.conf:app-editors/vim ltofat.conf
```

I have also just tried updating binutils, which **didn’t** help. The errors are still the same. But then I’ve changed USE flags… and voilà! It builds just fine with default flags.
I’ll pin-point the exact flag that is the culprit and will make a PR with appropriate comment.

**Comment by InBetweenNames** [comment]
Got it, thanks for investigating!  I actually don't have the perl USE flag either.  In fact, I should probably place a special note in the README about Perl 5... even the developers themselves claim it will likely never work with LTO.
